sql/memo/rel_props.go

```diff
@@ -263,64 +260,64 @@ func (p *relProps) populateFds() {
 	p.fds = fds
 }
 
+func CardMemoGroups(g *ExprGroup) {
+	if g.RelProps.stat != nil {
+		return
+	}
+	// lookup joins and index scans will be first, if any
+	for _, g := range g.children() {
+		CardMemoGroups(g)
+	}
+	s := statsForRel(g.First)
+	g.RelProps.SetStats(s)
+}
+
 func statsForRel(rel RelExpr) sql.Statistic {
 	var stat sql.Statistic
 	switch rel := rel.(type) {
 	case JoinRel:
 		// different joins use different ways to estimate cardinality of outputs
 		jp := rel.JoinPrivate()
-		left := float64(jp.Left.RelProps.GetStats().RowCount())
-		right := float64(jp.Left.RelProps.GetStats().RowCount())
+		left := jp.Left.RelProps.GetStats()
+		right := jp.Left.RelProps.GetStats()
 
 		switch n := rel.(type) {
 		case *LookupJoin:
-			var card float64
 			if n.Injective {
-				card = n.Left.RelProps.card
-			} else {
-				sel := lookupJoinSelectivity(n.Lookup, n.JoinBase)
-				card = n.Left.RelProps.card * n.Right.RelProps.card * sel
+				return left
 			}
-			return &stats.Statistic{RowCnt: uint64(card)}
-		case *ConcatJoin:
-			var sel float64
-			for _, l := range n.Concat {
-				lookup := l
-				sel += lookupJoinSelectivity(lookup, n.JoinBase)
-			}
-			card := n.Left.RelProps.card * optimisticJoinSel * sel
-			return &stats.Statistic{RowCnt: uint64(card)}
-		case *LateralJoin:
-			card := n.Left.RelProps.card * n.Right.RelProps.card
-			return &stats.Statistic{RowCnt: uint64(card)}
 		default:
 		}
 
-		switch {
-		case jp.Op.IsPartial():
-			return &stats.Statistic{RowCnt: uint64(left * right * optimisticJoinSel)}
-		case jp.Op.IsLeftOuter():
-			card := math.Max(jp.Left.RelProps.card, optimisticJoinSel*left*right)
-			return &stats.Statistic{RowCnt: uint64(card)}
-		case jp.Op.IsRightOuter():
-			card := math.Max(jp.Right.RelProps.card, optimisticJoinSel*jp.Left.RelProps.card*jp.Right.RelProps.card)
-			return &stats.Statistic{RowCnt: uint64(card)}
-		case jp.Op.IsDegenerate():
-			return &stats.Statistic{RowCnt: uint64(left * right)}
-		default:
-			card := optimisticJoinSel * left * right
-			return &stats.Statistic{RowCnt: uint64(card)}
+		distinct := left.DistinctCount()
+		if right.DistinctCount() > distinct {
+			distinct = right.DistinctCount()
+		}
+		if distinct == 0 {
+			m := float64(left.RowCount())
+			if cmp := float64(left.RowCount()); cmp > m {
+				m = cmp
+			}
+			distinct = uint64(m * .80)
 		}
+		card := float64(left.RowCount()*right.RowCount()) / float64(distinct)
```

Comment:
Is `distinct` the number of distinct rows in the node? It's not clear what the intent behind this computation is; maybe it could use a comment.